tasks/Task.cpp

```diff
@@ -2,48 +2,525 @@
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+            {
+                Json::FastWriter fast;
+                ws->send(fast.write(message));
+            }
+        });
+
+    return peer_connection;
 }
 
-Task::~Task()
+void Task::configurePeerDataChannel()
 {
+    mPeerConnection->onDataChannel(
+        [&](shared_ptr<rtc::DataChannel> data_channel)
+        {
+            mDataChannel = data_channel;
+            registerDataChannelCallBacks(data_channel);
+        });
 }
 
+void Task::configureWebSocket()
+{
+    promise<void> ws_promise;
+    future<void> ws_future = ws_promise.get_future();
+
+    mWs->onOpen(
+        [&]()
+        {
+            LOG_INFO_S << "WebSocket connected, signaling ready" << endl;
+            mState.web_socket = WebSocketOpened;
+            ws_promise.set_value();
+        });
+
+    mWs->onError(
+        [&](const string &error)
+        {
+            LOG_ERROR_S << "WebSocket failed: " << error << endl;
+            mState.web_socket = WebSocketFailed;
+            ws_promise.set_exception(make_exception_ptr(runtime_error(error)));
+            mPeerConnectionClosePromise.set_exception(make_exception_ptr(runtime_error(error)));
+            trigger();
+        });
+
+    mWs->onClosed(
+        [&]()
+        {
+            LOG_INFO_S << "WebSocket closed" << endl;
+            mState.web_socket = WebSocketClosed;
+            trigger();
+        });
+
+    mWs->onMessage(
+        [&](variant<binary, string> data)
+        {
+            if (!holds_alternative<string>(data))
+            {
+                return;
+            }
 
+            parseIncomingMessage(get<string>(data).c_str());
+            string actiontype = mDecoder.getActionType();
+
+            if (actiontype == "ping")
+            {
+                pong();
+            }
+            if (actiontype == "pong")
+            {
+                if (_remote_peer_id.get() == mDecoder.getFrom())
+                {
+                    mWaitRemotePeerPromise.set_value();
+                }
+                else
+                {
+                    mWaitRemotePeerPromise.set_exception(
+                        make_exception_ptr(runtime_error("Remote peer unreachable")));
+                }
+            }
+
+            if (actiontype == "offer")
+            {
+                onOffer();
+            }
+            else if (actiontype == "answer")
+            {
+                onAnswer();
+            }
+            else if (actiontype == "candidate")
+            {
+                onCandidate();
+            }
+        });
+
+    // wss://signalserverhost?user=yourname
+    const string url = _signaling_server_name.get() + "?user=" + _local_peer_id.get();
+    mWs->open(url);
+    future_status status =
+        ws_future.wait_until(chrono::system_clock::now() +
+                             chrono::seconds(static_cast<int>(_websocket_open_time_out.get().toSeconds())));
```

Comment:
`_websocket_open_time_out` is a `base::Time`. It has a microsecond resolution. Converting to seconds like this makes sure that one day someone will pick a sub-second value and things will break. Use microseconds.